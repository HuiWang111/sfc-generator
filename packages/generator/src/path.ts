import type { ElementNode, TemplateNode } from './types'
import { isElementNode } from './is'

export class Path {
  constructor(private _node: TemplateNode) {}

  get parent() {
    return this._node.parent
  }

  get node() {
    return this._node
  }

  private _find(
    node: TemplateNode,
    predicate: (node: TemplateNode) => boolean,
  ): TemplateNode | null {
    if (isElementNode(node) && node.children) {
      for (const child of node.children) {
        if (predicate(child)) {
          return child
        }
        else if (isElementNode(child)) {
          const result = this._find(child, predicate)
          if (result) {
            return result
          }
        }
      }
    }
    return null
  }

  replaceWith(node: TemplateNode) {
    const parent = this._node.parent as ElementNode | undefined
    if (parent && parent.children) {
      const index = parent.children.indexOf(this._node)
      if (index > -1) {
        parent.children[index] = node
      }
    }
    node.parent = parent
    this._node = node
  }

  remove() {
    const parent = this._node.parent as ElementNode | undefined
    if (parent && parent.children) {
      const index = parent.children.indexOf(this._node)
      if (index > -1) {
        parent.children.splice(index, 1)
      }
    }
  }

  skip() {
    if (isElementNode(this._node) && this._node.children) {
      this._node.children.forEach((child) => {
        child.__skip__ = true
      })
    }
  }

  getSibling(index: number) {
    const parent = this._node.parent as ElementNode | undefined
    if (parent && parent.children) {
      return parent.children[index]
    }
    return null
  }

  find(predicate: (node: TemplateNode) => boolean) {
    return this._find(this._node, predicate)
  }
}
