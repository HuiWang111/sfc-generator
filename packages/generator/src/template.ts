import type { ElementNode, GenerateOptions, TemplateNode } from './types'
import { TemplateNodeType } from './types'

export class Template {
  constructor(private options: GenerateOptions = {}) {}

  private padCode(str: string, indentCount: number) {
    const { indent = 2 } = this.options
    const spaceCount = indent * indentCount
    return str.padStart(str.length + spaceCount)
  }

  private generateAttrCode(
    node: ElementNode,
    indentCount: number,
  ): [string, boolean] {
    const { breakOnAttrs = 2 } = this.options
    const attrEntries = Object.entries(node.attrsMap)
    if (!attrEntries.length) {
      return ['', false]
    }

    const shouldBreak = attrEntries.length > breakOnAttrs
    const attrCodes = (
      attrEntries
        .map(([key, value]) => {
          if (shouldBreak) {
            return this.padCode(`${key}="${value}"`, indentCount)
          }
          return `${key}="${value}"`
        })
    )

    if (!shouldBreak) {
      return [attrCodes.join(' '), shouldBreak]
    }
    return [attrCodes.join('\n'), shouldBreak]
  }

  private generateEndTag(
    node: ElementNode,
    indentCount: number,
    shouldBreak: boolean,
  ) {
    if (!node.children || !node.children.length) {
      return this.padCode('/>', indentCount)
    }
    const childrenCode = (
      node.children
        .map((c) => {
          return this.generate(c, indentCount + 1)
        })
        .join('')
    )
    return [
      this.padCode('>', shouldBreak ? indentCount : 0),
      childrenCode,
      this.padCode(`</${node.tag}>`, indentCount),
    ].join('\n')
  }

  generate(node: TemplateNode, indentCount = 1): string {
    if (
      node.type === TemplateNodeType.Interpolation
      || node.type === TemplateNodeType.Text
    ) {
      if (node.type === TemplateNodeType.Text && node.text === ' ') {
        return '\n'
      }
      return this.padCode(node.text, indentCount)
    }

    const [attrCode, shouldBreak] = this.generateAttrCode(node, indentCount + 1)
    const startTag = this.padCode(`<${node.tag}`, indentCount)
    const endTag = this.generateEndTag(node, indentCount, shouldBreak)

    if (!shouldBreak) {
      return [
        startTag,
        attrCode ? ' ' : '',
        attrCode,
        endTag,
      ]
        .join('')
    }
    return [startTag, attrCode, endTag].join('\n')
  }
}
