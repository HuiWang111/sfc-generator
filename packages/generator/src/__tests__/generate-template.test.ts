import type { TemplateNode } from '../types'
import { readFile } from 'node:fs/promises'
import { join } from 'node:path'
import { compileTemplate, parseComponent } from '@vue/compiler-sfc'
import { Template } from '../template'

describe('test generate template', () => {
  const template = new Template()

  it('test generate simple node', () => {
    const res = compileTemplate({
      source: '<div>test</div>',
      filename: '',
    })

    const code = template.generate(res.ast as TemplateNode)
    expect(code).toMatchSnapshot()
  })

  it('test generate simple node with attrs', () => {
    const res = compileTemplate({
      source: '<div class="test" v-if="visible">test</div>',
      filename: '',
    })

    const code = template.generate(res.ast as TemplateNode)
    expect(code).toMatchSnapshot()
  })

  it('test generate simple node with interpolation expression', () => {
    const res = compileTemplate({
      source: '<div>{{ test }}</div>',
      filename: '',
    })

    const code = template.generate(res.ast as TemplateNode)
    expect(code).toMatchSnapshot()
  })

  it('test generate component node', () => {
    const res = compileTemplate({
      source: '<Test>test</Test>',
      filename: '',
    })

    const code = template.generate(res.ast as TemplateNode)
    expect(code).toMatchSnapshot()

    const res1 = compileTemplate({
      source: '<form-create>test</form-create>',
      filename: '',
    })

    const code1 = template.generate(res1.ast as TemplateNode)
    expect(code1).toMatchSnapshot()
  })

  it('test generate form component', async () => {
    const sfc = await readFile(join(__dirname, 'components/form.vue'), 'utf-8')
    const sfcDescriptor = parseComponent(sfc)

    if (sfcDescriptor.template) {
      const res = compileTemplate({
        source: sfcDescriptor.template.content,
        filename: '',
      })

      const code = template.generate(res.ast as TemplateNode)
      expect(code).toMatchSnapshot()
    }
  })

  it('test generate with scopedSlots', async () => {
    const source = `<n-selector style="width: 30px;">
  <span>children</span>
  <template #trigger="{ toggle }">
    <div id="portrait-more" style="cursor: pointer;width: 30px;height: 30px;" @click="toggle" v-html="moreIcon" />
  </template>
  <template #panel>
    <n-selector-item>
      <help label="帮助" :icon-size="26" :config-list="configList">
        <template #title>
          <div>自定义 title</div>
        </template>
      </help>
    </n-selector-item>
  </template>
</n-selector>`
    const res = compileTemplate({
      source,
      filename: '',
    })

    const code = template.generate(res.ast as TemplateNode)
    expect(code).toMatchSnapshot()
  })
})
