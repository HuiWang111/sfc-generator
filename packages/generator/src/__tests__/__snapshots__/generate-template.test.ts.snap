// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test generate template test generate component node 1`] = `
"  <Test>
    test
  </Test>"
`;

exports[`test generate template test generate component node 2`] = `
"  <form-create>
    test
  </form-create>"
`;

exports[`test generate template test generate form component 1`] = `
"  <form-creator-config-provider :locale="locale">
    <div class="container">
      <div class="submit-container">
        <a-button
          type="primary"
          @click="handleSubmit"
          class="submit-btn"
        >
          {{$t('saveBtn')}}
        </a-button>
      </div>
      <div
        v-if="a"
        v-show="b"
        v-for="data of formCreateData"
        :key="data.key"
        class="form-item"
        @click="handleSubmit"
      >
        <h2>
          {{ data.title }}
        </h2>
        <form-create
          v-model="formApis[data.key]"
          :rule="data.formItems"
          :option="options"
        />
      </div>
    </div>
  </form-creator-config-provider>"
`;

exports[`test generate template test generate simple node 1`] = `
"  <div>
    test
  </div>"
`;

exports[`test generate template test generate simple node with attrs 1`] = `
"  <div class="test" v-if="visible">
    test
  </div>"
`;

exports[`test generate template test generate simple node with interpolation expression 1`] = `
"  <div>
    {{ test }}
  </div>"
`;

exports[`test generate template test generate with scopedSlots 1`] = `
"  <n-selector style="width: 30px;">
    <template #trigger="{ toggle }">
      <div
        id="portrait-more"
        style="cursor: pointer;width: 30px;height: 30px;"
        @click="toggle"
        v-html="moreIcon"
      />
    </template>
    <template #panel>
      <n-selector-item>
        <help
          label="帮助"
          :icon-size="26"
          :config-list="configList"
        >
          <template #title>
            <div>
              自定义 title
            </div>
          </template>
        </help>
      </n-selector-item>
    </template>
    <span>
      children
    </span>
  </n-selector>"
`;
